### PR TITLE
Cryptocurrency conversion feature [US #108] [Task #130 #131 #132]

### PR DESCRIPTION
**Pull request by miisaka**
PR for user story #108, task #130 #131 #132

### WHAT kind of change does this PR introduce?
Added a new view for the cryptocurrency conversion feature that supports conversion between the major cryptocurrencies Bitcoin, Ethereum, Litecoin, Ripple and USD
...

### HOW is this accomplished?
It is done by Implementing a new view that shows 6 decimal numbers and that will convert between BTC, ETH, LTC, XRP and USD by grabbing live data from https://min-api.cryptocompare.com through an asynchronous task. 

ezgif-1-a93c135758

...

### Checklist
- [x] Docs have been added / updated to reflect the changes
- [x] I have reviewed and tested the changes :white_check_mark:
- [x] I have squashed my commits to have a reasonable amount of commits
- [x] All of the code I have written adhere to the coding conventions outlined in the coding conventions section of the wiki

### CODE REVIEW

**Comment by izconcept** [comment]
Can you amend the last commit so that it's linked to an issue?

File: app/src/androidTest/java/com/simplemobiletools/calculator/CryptoActivityTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        longPress((R.id.btn_delete));
+        checkResult("0.000000");
+        press(R.id.btn_delete);
+        checkResult("0.000000");
+    }
+
+    @Test
+    public void convertBTCtoETH() {
+        BackgroundCryptoTaskBuilder taskBuilder = new BackgroundCryptoTaskBuilder(this.activity.getActivity(), this.activity.getActivity().calc);
+        BackgroundCryptoTaskBuilder fakeBuilder = spy(taskBuilder);
```
Comment (almiche) [inline app/src/androidTest/java/com/simplemobiletools/calculator/CryptoActivityTest.java]: Nice solution c:

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        // Sets whether the selector wheel wraps when reaching the min/max value.
+        numberPickerFROM.wrapSelectorWheel = true
+        numberPickerTO.wrapSelectorWheel = true
+
+        val cryptoSymbolHashMap = HashMap<String, String>()
+        cryptoSymbolHashMap.put("Ethereum", "ETH")
+        cryptoSymbolHashMap.put("Bitcoin", "BTC")
+        cryptoSymbolHashMap.put("Ripple", "XRP")
+        cryptoSymbolHashMap.put("Litecoin", "LTC")
+        cryptoSymbolHashMap.put("USD", "USD")
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: No bitconnect 🧐

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                calc.numpadClicked(it.id); checkHaptic(it)
+            }
+        }
+
+        var numberPickerFROM: NumberPicker = numberPicker1
+        var numberPickerTO: NumberPicker = numberPicker2
+
+        // Values of each of the Number Pickers
+        val valuesFROM = arrayOf("Bitcoin", "Ethereum", "Ripple", "Litecoin", "USD")
+        val valuesTO = arrayOf("Ethereum", "Ripple", "Litecoin", "USD", "Bitcoin")
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: I believe these can be set inside of arrays.xml in order to be more easily maintained [Documentation for reference](https://developer.android.com/samples/MediaRouter/res/values/arrays.html)

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        result.setOnLongClickListener { copyToClipboard(result.value); true }
+
+        AutofitHelper.create(result)
+    }
+
+    private fun performConversion(cryptoFROM: String, cryptoTO: String) {
+        if (isOnline()) {
+            calc.calculateCrypto(cryptoFROM, cryptoTO, this)
+        } else if (!isOnline()) {
+            displayToast("It seems there has been a connection problem, contact you ISP for more details !")
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: Maybe we should consider displaying an alertDialog to the user. This toast is too easy to dismiss. Perhaps something similar using the code for my error handling feature.

Comment (almiche) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: [PR for reference](https://github.com/jusleg/simple-calculator-FAMINGO/pull/155)

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    override fun onResume() {
+        super.onResume()
+        if (storedUseEnglish != config.useEnglish) {
+            restartActivity()
+            return
+        }
+
+        if (storedTextColor != config.textColor) {
+            updateViewColors(crypto_holder, config.textColor)
+//            updateButtonColor(config.customPrimaryColor)
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: Is this necessary ?

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 6]
```diff
@@ -162,6 +164,9 @@ class MoneyActivity : SimpleActivity(), Calculator, LocationListener {
                     spawnTaxModal()
                     return
                 }
+                var taxRate: Double = (TaxOperation.getTaxRate(province)) * 100
+                var message: String = "Current Location: " + province + "\nTax rate: " + taxRate + "% "
+                Toast.makeText(applicationContext, message, Toast.LENGTH_LONG).show()
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Is this part of crypto conversion ?

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            number = number.dropLast(1 + decimalCounter)
+            decimalClicked = false
+            decimalCounter = 0
+        } else if (decimalClicked) {
+            decimalCounter--
+            number = number.dropLast(1)
+        } else {
+            number = number.dropLast(1)
+        }
+        setValue(String.format("%,.6f", number.toDouble()))
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt]: why not use overwriteNumber here ?

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    private fun setValue(value: String) {
+        mCallback!!.setValue(value)
+    }
+
+    private fun getResult() = Formatter.stringToDouble(mCallback!!.getResult())
+
+    fun calculateCrypto2(from: String, to: String) {
+        var response: AsyncTask<Void, Void, String>? = taskBuilder.from(from).to(to).build().execute()
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt]: Why is the type of this not BackgroundCryptoTaskBuilder and furthermore why is there a question mark signalling thats it is nullable ?

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    fun supersedeBuilder(builder: BackgroundCryptoTaskBuilder) {
+        taskBuilder = builder
+        testing = true
+    }
+
+    fun calculateCrypto(cryptoFROM: String, cryptoTO: String, cryptoActivity: CryptoActivity) {
+        if (testing) {
+            calculateCrypto2(cryptoFROM, cryptoTO)
+        } else {
+            var response = GetCryptoTask(cryptoFROM, cryptoTO, cryptoActivity, this).execute()
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt]: I think you should be using the builder here to return a new GetCryptoTask or a FakeCryptoTask. Using polymorphism rather than this testing flag.

Comment (almiche) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt]: Contact me for more info xD

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/GetCryptoTask.kt [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    var cryptoCalculator: CryptoCalculatorImpl = cryptoCalculatorImpl
+
+    override fun doInBackground(vararg params: Void?): String? {
+        // old API that crashes after too many requests, keeping
+        //val url = URL("https://api.cryptonator.com/api/ticker/$cryptoFROM-$cryptoTO")
+        val url = URL("https://min-api.cryptocompare.com/data/price?fsym=$cryptoFROM&tsyms=$cryptoTO")
+
+        val httpClient = url.openConnection() as HttpURLConnection
+
+        if (httpClient.responseCode == HttpURLConnection.HTTP_OK) {
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/GetCryptoTask.kt]: Error handling should go here add a switch case for HttpURLConnection.HTTP_NOT_FOUND and trigger an alert dialog in the ui thread.

File: commons/src/main/java/com/simpletools/calculator/commons/operations/TaxOperation.kt [hunk 11]
```diff
@@ -35,4 +35,8 @@ object TaxOperation {
     fun getFormula(): String {
         return "Tax in" + location + " is " + taxesMap[location]
     }
+
+    fun getTaxRate(province: String): Double {
+        return taxesMap[province]!!
```
Comment (almiche) [inline commons/src/main/java/com/simpletools/calculator/commons/operations/TaxOperation.kt]: What is this doing ?

File: testing [hunk 12]
```diff
@@ -1,6 +1,6 @@
 if [[ $1 == integration_tests ]]; then
 	echo no | android create avd --force -n test -t android-22 --abi armeabi-v7a
-	emulator -avd test -no-skin -no-audio -no-window &
+	emulator -avd test -no-audio -no-window &
```
Comment (almiche) [inline testing]: Nice catch no skin is deprecated I believe

**Comment by almiche** [comment]
✅ Styling is adequate and follows coding conventions outlined by the [Kotlin Foundation](https://kotlinlang.org/docs/reference/coding-conventions.html)

✅ New methods and functionalities have appropriate tests (testing edge cases + good coverage)
** Testing is through **

✅ Branch has an acceptable amount of commits with descriptive names and

✅   PR description matches the code outlined a has mentions to issues

✅ This PR attempts to implement a specific feature (no code blob or epic pr)

** Overall this is good there a few lines of code pertaining to tax conversions if i'm not mistaken. **

Status: **Requested Changes**

**Comment by jusleg** [comment]
@miisaka first commit name should be changed using an interactive rebase. Look at the last answer on [this post](https://stackoverflow.com/questions/1884474/change-old-commit-message-on-git?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa)

**Comment by miisaka** [comment]
@izconcept @jusleg changed the name of the Tax toast commit by linking it to issue #140.

File: app/src/androidTest/java/com/simplemobiletools/calculator/CryptoActivityTest.java (hunk 1 above)
Comment (miisaka) [reply app/src/androidTest/java/com/simplemobiletools/calculator/CryptoActivityTest.java]: Thank you!

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt (hunk 2 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: Unfortunately our application does not support scammers

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt (hunk 3 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: I shall implement this refactoring. Thank you for the suggestion kind sir

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt (hunk 5 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: It is not. I shall remove that commented line, thank you for spotting it.

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 6 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: It is not, I have modified the commit message to link it to issue #140

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt (hunk 7 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt]: Because overwriteNumber sets decimalClicked to false and decimalCounter to 0 which will reset the decimals on the value text field if the user had entered decimals before pressing on the backspace button.

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt (hunk 9 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt]: Indeed this could be refactored to be cleaner. I shall implement it.

File: commons/src/main/java/com/simpletools/calculator/commons/operations/TaxOperation.kt (hunk 11 above)
Comment (miisaka) [reply commons/src/main/java/com/simpletools/calculator/commons/operations/TaxOperation.kt]: It is used to display the tax rate on a Toast to the user upon pressing on the tax button.

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/GetCryptoTask.kt (hunk 10 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/GetCryptoTask.kt]: Will add more error handling for this.

File: app/src/androidTest/java/com/simplemobiletools/calculator/FakeCryptoTask.kt [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+import org.json.JSONException
+import org.json.JSONObject
+
+class FakeCryptoTask(val cryptoFROM: String, val cryptoTO: String, cryptoActivity: CryptoActivity, cryptoCalculatorImpl: CryptoCalculatorImpl) : AsyncTask<Void, Void, String>() {
+
+    var mCalc: Calculator = cryptoActivity
+    var cryptoCalculator: CryptoCalculatorImpl = cryptoCalculatorImpl
+
+    override fun doInBackground(vararg params: Void?): String? {
+        // old API stub
```
Comment (jusleg) [inline app/src/androidTest/java/com/simplemobiletools/calculator/FakeCryptoTask.kt]: You should not have commented code

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+            }
+        }
+
+        var numberPickerFROM: NumberPicker = numberPicker1
+        var numberPickerTO: NumberPicker = numberPicker2
+
+        // Values of each of the Number Pickers
+        val valuesFROM = resources.getStringArray(R.array.from_crypto_array)
+        val valuesTO = resources.getStringArray(R.array.to_crypto_array)
+//        val valuesFROM = arrayOf("Bitcoin", "Ethereum", "Ripple", "Litecoin", "USD")
```
Comment (jusleg) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: should remove if not used anymore

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/GetCryptoTask.kt [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+import java.net.URL
+
+class GetCryptoTask(val cryptoFROM: String, val cryptoTO: String, cryptoActivity: CryptoActivity, cryptoCalculatorImpl: CryptoCalculatorImpl) : AsyncTask<Void, Void, String>() {
+
+    private val cryptoActivity: CryptoActivity = cryptoActivity
+    private var cryptoCalculator: CryptoCalculatorImpl = cryptoCalculatorImpl
+
+    override fun doInBackground(vararg params: Void?): String? {
+        try {
+            // old API that crashes after too many requests, keeping
```
Comment (jusleg) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/GetCryptoTask.kt]: comments... need to go

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt (hunk 2 above)
Comment (almiche) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: :'(

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt (hunk 5 above)
Comment (almiche) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/CryptoActivity.kt]: c:

**Comment by almiche** [comment]
✅ Styling is adequate and follows coding conventions outlined by the [Kotlin Foundation](https://kotlinlang.org/docs/reference/coding-conventions.html)

✅ New methods and functionalities have appropriate tests (testing edge cases + good coverage)

**Testing is through**

✅ Branch has an acceptable amount of commits with descriptive names and

✅   PR description matches the code outlined a has mentions to issues

✅ This PR attempts to implement a specific feature (no code blob or epic pr)

**Overall this is good there a few lines of code pertaining to tax conversions if i'm not mistaken.**

Status: **Approved**

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/GetCryptoTask.kt [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+            httpClient.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
+            httpClient.setRequestProperty("charset", "utf-8")
+            httpClient.useCaches = false
+
+            if (httpClient.responseCode == HttpURLConnection.HTTP_OK) {
+                val stream = BufferedInputStream(httpClient.inputStream)
+                val data: String = readStream(inputStream = stream)
+                httpClient.disconnect()
+                return data
+            } else if (httpClient.responseCode == HttpURLConnection.HTTP_NOT_FOUND) {
```
Comment (jusleg) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/GetCryptoTask.kt]: @miisaka  I would delete all those else if and only keep the global else. You are basically doing the same thing in every `else if`

Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/GetCryptoTask.kt]: I have removed them to improve code quality.
